### PR TITLE
[ML][Pipelines]feat: support component download

### PR DESCRIPTION
**Pull request by elliotzh**
# Description

add a new experimental public API for component operations: `ml_client.components.download`.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py [hunk 1]
```diff
@@ -217,6 +220,43 @@ def get(self, name: str, version: Optional[str] = None, label: Optional[str] = N
         :return: The specified component object.
         :rtype: ~azure.ai.ml.entities.Component
         """
+        return self._get(name=name, version=version, label=label)
+
+    @experimental
+    @monitor_with_telemetry_mixin(logger, "Component.Get", ActivityType.PUBLICAPI)
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py:226]: Get -> Download

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py [hunk 2]
```diff
@@ -217,6 +220,43 @@ def get(self, name: str, version: Optional[str] = None, label: Optional[str] = N
...
+    def download(
+        self,
+        name: str,
+        output_dir: Union[str, Path, PathLike],
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py:230]: Let's follow existing API design, like model download

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py [hunk 3]
```diff
@@ -217,6 +220,43 @@ def get(self, name: str, version: Optional[str] = None, label: Optional[str] = N
...
+        *,
+        version: Optional[str] = None,
+        label: Optional[str] = None
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py:233]: hide label? seems v2 still not fully supported label

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py [hunk 4]
```diff
@@ -217,6 +220,43 @@ def get(self, name: str, version: Optional[str] = None, label: Optional[str] = N
...
+    ) -> Component:
+        """Returns information about the specified component.
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py:235]: update docstring

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py (hunk 2 above)
Comment (elliotzh) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py]: move `version` behind `download_path` as download component without `version` provided is possible.

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_component/component.py [hunk 5]
```diff
@@ -521,6 +521,12 @@ def _to_dict(self) -> Dict:
         # TODO: handle other_parameters and remove override from subclass
         return component_schema_dict
 
+    def _localize(self):
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_component/component.py:524]: add some docstring on what does _localize means

File: sdk/ml/azure-ai-ml/tests/component/e2etests/test_component.py [hunk 6]
```diff
@@ -1079,3 +1080,28 @@ def test_create_component_skip_if_no_change(self, client: MLClient, randstr):
...
+        component.name = randstr("component_name")
+        _ = client.components.create_or_update(
+            component,
+        )
+
+        output_dir = f"./tests/test_configs/components/downloaded/{request.node.callspec.id}"
+        if os.path.isdir(output_dir):
+            shutil.rmtree(Path(output_dir))
+
+        client.components.download(
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/tests/component/e2etests/test_component.py:1103]: assert on the downloaded code snapshot?

File: sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py [hunk 7]
```diff
@@ -63,31 +63,51 @@ def _get_datastore_name(*, datastore_name: Optional[str] = WORKSPACE_BLOB_STORE)
     return datastore_name
 
 
-def get_datastore_info(operations: DatastoreOperations, name: str) -> Dict[str, str]:
-    """Get datastore account, type, and auth information."""
+def get_datastore_info(
+    operations: DatastoreOperations,
+    name: str,
+    *,
+    local_credential=None,
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py:70]: maybe just credential ?

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_code_operations.py [hunk 8]
```diff
@@ -192,6 +202,67 @@ def get(self, name: str, version: str) -> Code:
...
+                name=_get_datastore_name(),
+                # local_credential=self._service_client._config.credential
+                container_name=m.group("container_name"),
+            )
+        except HttpResponseError:
+            datastore_info = get_datastore_info(
+                self._datastore_operation,
+                # always use WORKSPACE_BLOB_STORE
+                name=_get_datastore_name(),
+                local_credential=self._service_client._config.credential,
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_code_operations.py:254]: shall this be the default path? As user should normally has permission to the default datastore?

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_code_operations.py [hunk 9]
```diff
@@ -192,6 +202,67 @@ def get(self, name: str, version: str) -> Code:
...
+                container_name=m.group("container_name"),
+            )
+        storage_client = get_storage_client(**datastore_info)
+        storage_client.download(
+            starts_with=m.group("blob_name"),
+            destination=output_dir.as_posix(),
+        )
+        if not output_dir.is_dir() or not output_dir.glob("**/*"):
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_code_operations.py:262]: output_dir.glob("**/*") this looks a bit strange to test dir empty, 
Maybe something like below is more straight forward.
len(list_dir())==0

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_code_operations.py (hunk 8 above)
Comment (elliotzh) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_code_operations.py:254]: As Alex has said, current solution actually depends on their backend implementation of snapshots. Current default path is used by most artifact upload and all artifact download paths in DPv2 is more likely to be well informed and maintained if there are any changes.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_code_operations.py [hunk 10]
```diff
@@ -225,26 +225,42 @@ def download(self, name: str, version: str, download_path: Union[PathLike, str])
...
-            credential=datastore_info["credential"],
-            storage_account=datastore_info["storage_account"],
-            container_name=m.group("container_name"),
-        )
+        try:
+            datastore_info = get_datastore_info(
+                self._datastore_operation,
+                # always use WORKSPACE_BLOB_STORE
+                name=_get_datastore_name(),
+                # local_credential=self._service_client._config.credential
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_code_operations.py:246]: remove this line

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**